features/mbedtls/platform/src/platform_alt.c

```diff
@@ -0,0 +1,59 @@
+ /*
+  *  platform_alt.c
+  *
+  *  Copyright (C) 2018, Arm Limited, All Rights Reserved
+  *  SPDX-License-Identifier: Apache-2.0
+  *
+  *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+  *  not use this file except in compliance with the License.
+  *  You may obtain a copy of the License at
+  *
+  *  http://www.apache.org/licenses/LICENSE-2.0
+  *
+  *  Unless required by applicable law or agreed to in writing, software
+  *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+  *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  *  See the License for the specific language governing permissions and
+  *  limitations under the License.
+  *
+  */
+
+#include "mbedtls/platform.h"
+#if defined(MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT)
+
+static int reference_count = 0;
+
+int mbedtls_platform_setup( mbedtls_platform_context *ctx )
+{
+    int ret = 0;
+    if( ctx == NULL )
+        return ( -1 );
+
+    reference_count++;
+
+    if( reference_count == 1 )
+    {
+        /* call platform specific code to setup crypto driver*/
+        ret = crypto_platform_setup( &ctx->platform_impl_ctx );
+    }
+    return ( ret );
+}
+
+void mbedtls_platform_teardown( mbedtls_platform_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    if( reference_count == 0 )
```

Comment:
I don't think you need two `if` statements, and it looks a bit odd.

Have you considered using a less than operator? eg.
```
    reference_count--;
    if( reference_count <= 0 )
    {
        crypto_platform_teardown();
    }
```

Reply:
I agree it looks a bit odd, but what you are suggesting is prone to errors:
1. `crypto_platform_teardown()` *might* be called twice, if `reference_count` is 0 or -1.
2. corner cases where we get the reference counter to -2, and then when some chooses to setup, it won't initialize the platform, because the counter will be increased to -1.

Perhaps assigning `reference_count` to 0 after calling `crypto_platform_teardown()`?

Reply:
I agree, it would be safer to reassign `reference_count`.

So...
```
    reference_count--;
    if( reference_count <= 0 )
    {
        crypto_platform_teardown();
        reference_count = 0;
    }
```